components/de_amp/browser/de_amp_url_loader.cc

```diff
@@ -0,0 +1,408 @@
+/* Copyright 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/de_amp/browser/de_amp_url_loader.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/bind.h"
+#include "base/metrics/histogram_macros.h"
+#include "base/no_destructor.h"
+#include "base/task/post_task.h"
+#include "base/task/thread_pool.h"
+#include "brave/components/de_amp/browser/de_amp_throttle.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/self_owned_receiver.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/mojom/early_hints.mojom.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+#include "third_party/re2/src/re2/re2.h"
+
+namespace de_amp {
+
+namespace {
+
+constexpr uint32_t kReadBufferSize = 65536;
+static const char kDetectAmpPattern[] = "(?:<.*html\\s.*(amp|⚡)\\s.*>)";
```

Comment:
Regular expressions are both fun and notoriously hard to read :)

I'd suggest a unit test to ensure that the regexps are working as expected.

Reply:
@fmarier Added unit tests for regexes: https://github.com/brave/brave-core/pull/11750/files#diff-d1ac8bd2af07faa8c34f994c9e5b86b2e67dc936f6c6a3d3b31e4965867ddc9d